<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (17) on Fri Jan 19 16:08:00 UTC 2024 -->
<title>Gene2Genotype (LIRICAL 2.0.0 API)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="dc.created" content="2024-01-19">
<meta name="description" content="declaration: package: org.monarchinitiative.lirical.core.model, interface: Gene2Genotype">
<meta name="generator" content="javadoc/ClassWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../../../script.js"></script>
<script type="text/javascript" src="../../../../../script-dir/jquery-3.6.1.min.js"></script>
<script type="text/javascript" src="../../../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="class-declaration-page">
<script type="text/javascript">var evenRowColor = "even-row-color";
var oddRowColor = "odd-row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../../../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top">
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../../../../index.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="nav-bar-cell1-rev">Class</li>
<li><a href="class-use/Gene2Genotype.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../help-doc.html#class">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li>Constr&nbsp;|&nbsp;</li>
<li><a href="#method-summary">Method</a></li>
</ul>
<ul class="sub-nav-list">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li>Constr&nbsp;|&nbsp;</li>
<li><a href="#method-detail">Method</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" value="search" disabled="disabled">
<input type="reset" id="reset-button" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="sub-title"><span class="package-label-in-type">Package</span>&nbsp;<a href="package-summary.html">org.monarchinitiative.lirical.core.model</a></div>
<h1 title="Interface Gene2Genotype" class="title">Interface Gene2Genotype</h1>
</div>
<section class="class-description" id="class-description">
<hr>
<div class="type-signature"><span class="modifiers">public interface </span><span class="element-name type-name-label">Gene2Genotype</span></div>
<div class="block"><a href="Gene2Genotype.html" title="interface in org.monarchinitiative.lirical.core.model">Gene2Genotype</a> represents variants that have been annotated to a single gene. The gene data includes
 the identifier of a gene, the variants annotated with respect to the gene, and convenience methods for using
 in the <code>LIRICAL</code> algorithm.
 <p>
 Note, we only need the variants that passed the filtering for the analysis.</div>
</section>
<section class="summary">
<ul class="summary-list">
<!-- ========== METHOD SUMMARY =========== -->
<li>
<section class="method-summary" id="method-summary">
<h2>Method Summary</h2>
<div id="method-summary-table">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button id="method-summary-table-tab0" role="tab" aria-selected="true" aria-controls="method-summary-table.tabpanel" tabindex="0" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table', 3)" class="active-table-tab">All Methods</button><button id="method-summary-table-tab1" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab1', 3)" class="table-tab">Static Methods</button><button id="method-summary-table-tab2" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab2', 3)" class="table-tab">Instance Methods</button><button id="method-summary-table-tab3" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab3', 3)" class="table-tab">Abstract Methods</button><button id="method-summary-table-tab5" role="tab" aria-selected="false" aria-controls="method-summary-table.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('method-summary-table', 'method-summary-table-tab5', 3)" class="table-tab">Default Methods</button></div>
<div id="method-summary-table.tabpanel" role="tabpanel">
<div class="summary-table three-column-summary" aria-labelledby="method-summary-table-tab0">
<div class="table-header col-first">Modifier and Type</div>
<div class="table-header col-second">Method</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code>default int</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code><a href="#deleteriousAlleleCount(java.lang.String,float)" class="member-name-link">deleteriousAlleleCount</a><wbr>(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId,
 float&nbsp;pathogenicityThreshold)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5">
<div class="block">Get the count of alleles of predicted pathogenic/deleterious variants in the gene for the <code>sampleId</code>.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code>default int</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code><a href="#filteredOutVariantCount()" class="member-name-link">filteredOutVariantCount</a>()</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5">
<div class="block">Get the count of variants annotated to this gene which failed the filtering.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code>org.monarchinitiative.phenol.annotations.formats.GeneIdentifier</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code><a href="#geneId()" class="member-name-link">geneId</a>()</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3">
<div class="block">Get the credentials of the gene.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code>default double</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code><a href="#getSumOfPathBinScores(java.lang.String,float)" class="member-name-link">getSumOfPathBinScores</a><wbr>(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId,
 float&nbsp;pathogenicityThreshold)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5">&nbsp;</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code>default boolean</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code><a href="#hasVariants()" class="member-name-link">hasVariants</a>()</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5">&nbsp;</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab1"><code>static <a href="Gene2Genotype.html" title="interface in org.monarchinitiative.lirical.core.model">Gene2Genotype</a></code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab1"><code><a href="#of(org.monarchinitiative.phenol.annotations.formats.GeneIdentifier,java.util.Collection,int)" class="member-name-link">of</a><wbr>(org.monarchinitiative.phenol.annotations.formats.GeneIdentifier&nbsp;geneId,
 <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html" title="class or interface in java.util" class="external-link">Collection</a>&lt;<a href="LiricalVariant.html" title="interface in org.monarchinitiative.lirical.core.model">LiricalVariant</a>&gt;&nbsp;passingVariants,
 int&nbsp;filteredOutVariantCount)</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab1">
<div class="block">Create <a href="Gene2Genotype.html" title="interface in org.monarchinitiative.lirical.core.model">Gene2Genotype</a> from provided data.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code>default int</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5"><code><a href="#pathogenicClinVarCount(java.lang.String)" class="member-name-link">pathogenicClinVarCount</a><wbr>(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId)</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab5">
<div class="block">Get the number of predicted pathogenic/deleterious alleles in the gene for the <code>sampleId</code>.</div>
</div>
<div class="col-first odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code>int</code></div>
<div class="col-second odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code><a href="#variantCount()" class="member-name-link">variantCount</a>()</code></div>
<div class="col-last odd-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3">
<div class="block">Get the count of variants annotated to this gene that passed the filtering.</div>
</div>
<div class="col-first even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code><a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html" title="class or interface in java.util.stream" class="external-link">Stream</a>&lt;<a href="LiricalVariant.html" title="interface in org.monarchinitiative.lirical.core.model">LiricalVariant</a>&gt;</code></div>
<div class="col-second even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3"><code><a href="#variants()" class="member-name-link">variants</a>()</code></div>
<div class="col-last even-row-color method-summary-table method-summary-table-tab2 method-summary-table-tab3">
<div class="block">Get a <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html" title="class or interface in java.util.stream" class="external-link">Stream</a> of variants annotated to this gene.</div>
</div>
</div>
</div>
</div>
</section>
</li>
</ul>
</section>
<section class="details">
<ul class="details-list">
<!-- ============ METHOD DETAIL ========== -->
<li>
<section class="method-details" id="method-detail">
<h2>Method Details</h2>
<ul class="member-list">
<li>
<section class="detail" id="of(org.monarchinitiative.phenol.annotations.formats.GeneIdentifier,java.util.Collection,int)">
<h3>of</h3>
<div class="member-signature"><span class="modifiers">static</span>&nbsp;<span class="return-type"><a href="Gene2Genotype.html" title="interface in org.monarchinitiative.lirical.core.model">Gene2Genotype</a></span>&nbsp;<span class="element-name">of</span><wbr><span class="parameters">(org.monarchinitiative.phenol.annotations.formats.GeneIdentifier&nbsp;geneId,
 <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html" title="class or interface in java.util" class="external-link">Collection</a>&lt;<a href="LiricalVariant.html" title="interface in org.monarchinitiative.lirical.core.model">LiricalVariant</a>&gt;&nbsp;passingVariants,
 int&nbsp;filteredOutVariantCount)</span></div>
<div class="block">Create <a href="Gene2Genotype.html" title="interface in org.monarchinitiative.lirical.core.model">Gene2Genotype</a> from provided data.</div>
<dl class="notes">
<dt>Parameters:</dt>
<dd><code>geneId</code> - the gene credentials.</dd>
<dd><code>passingVariants</code> - a collection of variants that passed the initial filtering.</dd>
<dd><code>filteredOutVariantCount</code> - the number of variants that failed the initial filtering.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="geneId()">
<h3>geneId</h3>
<div class="member-signature"><span class="return-type">org.monarchinitiative.phenol.annotations.formats.GeneIdentifier</span>&nbsp;<span class="element-name">geneId</span>()</div>
<div class="block">Get the credentials of the gene.</div>
</section>
</li>
<li>
<section class="detail" id="variants()">
<h3>variants</h3>
<div class="member-signature"><span class="return-type"><a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html" title="class or interface in java.util.stream" class="external-link">Stream</a>&lt;<a href="LiricalVariant.html" title="interface in org.monarchinitiative.lirical.core.model">LiricalVariant</a>&gt;</span>&nbsp;<span class="element-name">variants</span>()</div>
<div class="block">Get a <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html" title="class or interface in java.util.stream" class="external-link">Stream</a> of variants annotated to this gene.</div>
<dl class="notes">
<dt>Returns:</dt>
<dd>a stream of variants found in this gene.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="variantCount()">
<h3>variantCount</h3>
<div class="member-signature"><span class="return-type">int</span>&nbsp;<span class="element-name">variantCount</span>()</div>
<div class="block">Get the count of variants annotated to this gene that passed the filtering.</div>
</section>
</li>
<li>
<section class="detail" id="hasVariants()">
<h3>hasVariants</h3>
<div class="member-signature"><span class="modifiers">default</span>&nbsp;<span class="return-type">boolean</span>&nbsp;<span class="element-name">hasVariants</span>()</div>
<dl class="notes">
<dt>Returns:</dt>
<dd><code>true</code> if the gene is annotated with 1 or more variants that passed the filtering.</dd>
</dl>
</section>
</li>
<li>
<section class="detail" id="filteredOutVariantCount()">
<h3>filteredOutVariantCount</h3>
<div class="member-signature"><span class="modifiers">default</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">filteredOutVariantCount</span>()</div>
<div class="block">Get the count of variants annotated to this gene which failed the filtering.</div>
</section>
</li>
<li>
<section class="detail" id="pathogenicClinVarCount(java.lang.String)">
<h3>pathogenicClinVarCount</h3>
<div class="member-signature"><span class="modifiers">default</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">pathogenicClinVarCount</span><wbr><span class="parameters">(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId)</span></div>
<div class="block">Get the number of predicted pathogenic/deleterious alleles in the gene for the <code>sampleId</code>.
 <p>
 Note, only the variant that passed the filtering are considered.</div>
</section>
</li>
<li>
<section class="detail" id="deleteriousAlleleCount(java.lang.String,float)">
<h3>deleteriousAlleleCount</h3>
<div class="member-signature"><span class="modifiers">default</span>&nbsp;<span class="return-type">int</span>&nbsp;<span class="element-name">deleteriousAlleleCount</span><wbr><span class="parameters">(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId,
 float&nbsp;pathogenicityThreshold)</span></div>
<div class="block">Get the count of alleles of predicted pathogenic/deleterious variants in the gene for the <code>sampleId</code>.
 The variants that are both <em>not</em> labeled as benign or likely benign by ClinVar and have the
 <a href="VariantMetadata.html#pathogenicityScore()"><code>VariantMetadata.pathogenicityScore()</code></a> at or above the provided <code>pathogenicityThreshold</code>
 are deemed to be predicted pathogenic/deleterious.
 <p>
 Note, we take specific precautions to not clash with ClinVar variant interpretation and consider ClinVar benign
 or likely benign variants as deleterious.</div>
</section>
</li>
<li>
<section class="detail" id="getSumOfPathBinScores(java.lang.String,float)">
<h3>getSumOfPathBinScores</h3>
<div class="member-signature"><span class="modifiers">default</span>&nbsp;<span class="return-type">double</span>&nbsp;<span class="element-name">getSumOfPathBinScores</span><wbr><span class="parameters">(<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html" title="class or interface in java.lang" class="external-link">String</a>&nbsp;sampleId,
 float&nbsp;pathogenicityThreshold)</span></div>
</section>
</li>
</ul>
</section>
</li>
</ul>
</section>
<!-- ========= END OF CLASS DATA ========= -->
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>Copyright &#169; 2024. All rights reserved.</small></p>
</footer>
</div>
</div>
</body>
</html>
