<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Running LIRICAL &mdash; LIRICAL 2.0.0-RC2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Output files" href="output.html" />
    <link rel="prev" title="Tutorial" href="tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            LIRICAL
          </a>
              <div class="version">
                2.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up LIRICAL</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running LIRICAL</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#shared-cli-options">Shared CLI options</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#resource-paths">Resource paths</a></li>
<li class="toctree-l3"><a class="reference internal" href="#configuration-options">Configuration options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-options">Output options</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#lirical-prioritization-commands">LIRICAL prioritization commands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#prioritize-run-lirical-with-via-cli-options"><code class="docutils literal notranslate"><span class="pre">prioritize</span></code> - run LIRICAL with via CLI options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#phenopacket-run-lirical-with-a-phenopacket"><code class="docutils literal notranslate"><span class="pre">phenopacket</span></code> - run LIRICAL with a Phenopacket</a></li>
<li class="toctree-l3"><a class="reference internal" href="#yaml-running-lirical-with-a-yaml-file"><code class="docutils literal notranslate"><span class="pre">yaml</span></code> - running LIRICAL with a YAML file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#choosing-between-yaml-and-phenopacket-input-formats">Choosing between YAML and Phenopacket input formats</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output files</a></li>
<li class="toctree-l1"><a class="reference internal" href="explanations.html">LIRICAL: How does it work?</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced.html">LIRICAL Advanced Options</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">LIRICAL</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Running LIRICAL</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/running.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="running-lirical">
<span id="rstrunning"></span><h1>Running LIRICAL<a class="headerlink" href="#running-lirical" title="Permalink to this heading"></a></h1>
<p>LIRICAL is a command-line Java tool that runs with Java version 17 or higher.
LIRICAL can be run both with and without genomic data in form of a VCF file from genome, exome,
or NGS gene-panel sequencing.</p>
<p>On typical computers, LIRICAL will run from about 15 to 60 seconds in phenotype-only mode,
~5 minutes with a typical exome file, or longer if a whole-genome file is used as input.</p>
<p>To get help, run LIRICAL with a command or with the option “-h”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="o">--</span><span class="n">help</span>
<span class="n">LIkelihood</span> <span class="n">Ratio</span> <span class="n">Interpretation</span> <span class="n">of</span> <span class="n">Clinical</span> <span class="n">AbnormaLities</span>

<span class="n">Usage</span><span class="p">:</span> <span class="n">lirical</span> <span class="p">[</span><span class="o">-</span><span class="n">hV</span><span class="p">]</span> <span class="p">[</span><span class="n">COMMAND</span><span class="p">]</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>      <span class="n">Show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
  <span class="o">-</span><span class="n">V</span><span class="p">,</span> <span class="o">--</span><span class="n">version</span>   <span class="n">Print</span> <span class="n">version</span> <span class="n">information</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
<span class="n">Commands</span><span class="p">:</span>
  <span class="n">download</span><span class="p">,</span> <span class="n">D</span>     <span class="n">Download</span> <span class="n">files</span> <span class="k">for</span> <span class="n">LIRICAL</span><span class="o">.</span>
  <span class="n">prioritize</span><span class="p">,</span> <span class="n">R</span>   <span class="n">Run</span> <span class="n">LIRICAL</span> <span class="kn">from</span> <span class="nn">CLI</span> <span class="n">arguments</span><span class="o">.</span>
  <span class="n">phenopacket</span><span class="p">,</span> <span class="n">P</span>  <span class="n">Run</span> <span class="n">LIRICAL</span> <span class="kn">from</span> <span class="nn">a</span> <span class="n">Phenopacket</span><span class="o">.</span>
  <span class="n">yaml</span><span class="p">,</span> <span class="n">Y</span>         <span class="n">Run</span> <span class="n">LIRICAL</span> <span class="kn">from</span> <span class="nn">a</span> <span class="n">YAML</span> <span class="n">file</span><span class="o">.</span>

<span class="n">See</span> <span class="n">the</span> <span class="n">full</span> <span class="n">documentation</span> <span class="n">at</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">lirical</span><span class="o">.</span><span class="n">readthedocs</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">en</span><span class="o">/</span><span class="n">master</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We assume that the <cite>lirical</cite> command alias was set as described in the <a class="reference internal" href="setup.html#rstsetupalias"><span class="std std-ref">Set up alias</span></a> section.</p>
</div>
<p>Run LIRICAL with a specific command with the <code class="docutils literal notranslate"><span class="pre">-h</span></code> option to get information about the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="n">download</span> <span class="o">-</span><span class="n">h</span>
<span class="n">Usage</span><span class="p">:</span> <span class="n">lirical</span> <span class="n">download</span> <span class="p">[</span><span class="o">-</span><span class="n">hVw</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="o">=&lt;</span><span class="n">datadir</span><span class="o">&gt;</span><span class="p">]</span>
<span class="n">Download</span> <span class="n">files</span> <span class="k">for</span> <span class="n">LIRICAL</span><span class="o">.</span>
  <span class="o">-</span><span class="n">d</span><span class="p">,</span> <span class="o">--</span><span class="n">data</span><span class="o">=&lt;</span><span class="n">datadir</span><span class="o">&gt;</span>   <span class="n">directory</span> <span class="n">to</span> <span class="n">download</span> <span class="n">data</span> <span class="p">(</span><span class="n">default</span><span class="p">:</span> <span class="n">data</span><span class="p">)</span>
  <span class="o">-</span><span class="n">w</span><span class="p">,</span> <span class="o">--</span><span class="n">overwrite</span>        <span class="n">overwrite</span> <span class="n">previously</span> <span class="n">downloaded</span> <span class="n">files</span> <span class="p">(</span><span class="n">default</span><span class="p">:</span> <span class="n">false</span><span class="p">)</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>             <span class="n">Show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
  <span class="o">-</span><span class="n">V</span><span class="p">,</span> <span class="o">--</span><span class="n">version</span>          <span class="n">Print</span> <span class="n">version</span> <span class="n">information</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
</pre></div>
</div>
<p>LIRICAL has four main commands, <code class="docutils literal notranslate"><span class="pre">download</span></code>, <code class="docutils literal notranslate"><span class="pre">prioritize</span></code>, <code class="docutils literal notranslate"><span class="pre">phenopacket</span></code>, and <code class="docutils literal notranslate"><span class="pre">yaml</span></code>.
We will <em>not</em> discuss the <code class="docutils literal notranslate"><span class="pre">download</span></code> command since it has already been covered in the <a class="reference internal" href="setup.html#rstdownload"><span class="std std-ref">LIRICAL data files</span></a> section</p>
<section id="shared-cli-options">
<h2>Shared CLI options<a class="headerlink" href="#shared-cli-options" title="Permalink to this heading"></a></h2>
<p>LIRICAL offers several commands for receiving phenotype and genotype inputs via CLI, phenopacket, or a YAML file.
However the commands share many CLI arguments for setting up the resource paths, the analysis configuration,
and where results should be written. We describe the shared CLI arguments in this section.</p>
<section id="resource-paths">
<h3>Resource paths<a class="headerlink" href="#resource-paths" title="Permalink to this heading"></a></h3>
<p>The options from this group point LIRICAL to resources required for analysis.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-d</span> <span class="pre">|</span> <span class="pre">--data</span></code>: path to LIRICAL data directory.
Required if the <code class="docutils literal notranslate"><span class="pre">data</span></code> folder is not set up next to the LIRICAL JAR file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-e19</span> <span class="pre">|</span> <span class="pre">--exomiser-hg19</span></code>: path to Exomiser variant database for <em>hg19</em>.
Required if the analysis is run with exome/genome sequencing files and <code class="docutils literal notranslate"><span class="pre">--assembly</span></code> is set to <em>hg19</em>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-e38</span> <span class="pre">|</span> <span class="pre">--exomiser-hg38</span></code>: path to Exomiser variant database for <em>hg38</em>.
Required if the analysis is run with exome/genome sequencing files and <code class="docutils literal notranslate"><span class="pre">--assembly</span></code> is set to <em>hg38</em>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-b</span> <span class="pre">|</span> <span class="pre">--background</span></code>: path to file with background variant frequencies for genes.
This option should not be used unless there is a very good reason to do that.
The background variant frequencies are bundled with the LIRICAL code. See <a class="reference internal" href="genotype-score.html#rstbg-var-freqs"><span class="std std-ref">Background variant frequencies</span></a> for more info.</p></li>
</ul>
</section>
<section id="configuration-options">
<h3>Configuration options<a class="headerlink" href="#configuration-options" title="Permalink to this heading"></a></h3>
<p>The configuration options tweak the analysis.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-g</span> <span class="pre">|</span> <span class="pre">--global</span></code>: global analysis, see <a class="reference internal" href="genotype-score.html#rstglobal-mode"><span class="std std-ref">Global mode</span></a> for more info (default: <code class="docutils literal notranslate"><span class="pre">false</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--ddndv</span></code>: disregard a disease if no deleterious variants are found in the gene associated with the disease.
Used only if running with a VCF file (default: <code class="docutils literal notranslate"><span class="pre">true</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--transcript-db</span></code>: transcript database (default: <code class="docutils literal notranslate"><span class="pre">RefSeq</span></code>), see <a class="reference internal" href="genotype-score.html#rsttx-dbs"><span class="std std-ref">Transcript databases</span></a> for more info.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--use-orphanet</span></code>: use <a class="reference external" href="https://www.orpha.net/consor/cgi-bin/index.php">Orphanet</a> annotations (default: <code class="docutils literal notranslate"><span class="pre">false</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--strict</span></code>: use strict penalties if the genotype does not match the disease model
in terms of number of called pathogenic alleles (default: <code class="docutils literal notranslate"><span class="pre">false</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--pathogenicity-threshold</span></code>: Variants with greater pathogenicity score is considered deleterious (default: <code class="docutils literal notranslate"><span class="pre">0.8</span></code>).</p></li>
</ul>
</section>
<section id="output-options">
<h3>Output options<a class="headerlink" href="#output-options" title="Permalink to this heading"></a></h3>
<p>The output options dictate the format and location for the analysis results.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">|</span> <span class="pre">--output-directory</span></code>: where to write the analysis outputs (default: current working directory).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-f</span> <span class="pre">|</span> <span class="pre">--output-format</span></code>: Output format to use for writing the results, can be provided multiple times.
Choose from <cite>html</cite>, <cite>tsv</cite>, and <cite>json</cite> (default: <code class="docutils literal notranslate"><span class="pre">html</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-x</span> <span class="pre">|</span> <span class="pre">--prefix</span></code>: prefix of the output files (default: <code class="docutils literal notranslate"><span class="pre">lirical</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-t</span> <span class="pre">|</span> <span class="pre">--threshold</span></code>: minimum post-test probability to show diagnosis in the HTML report.
The value must be in range <span class="math notranslate nohighlight">\([0, 1]\)</span>. The option must not be used with <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">|</span> <span class="pre">-mindiff</span></code> option at the same time.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">|</span> <span class="pre">--mindiff</span></code>: Minimal number of differential diagnoses to show.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--display-all-variants</span></code>: Display all variants in the HTML report, not just the variants passing
the pathogenicity threshold (default: <code class="docutils literal notranslate"><span class="pre">false</span></code>).</p></li>
</ul>
</section>
</section>
<section id="lirical-prioritization-commands">
<h2>LIRICAL prioritization commands<a class="headerlink" href="#lirical-prioritization-commands" title="Permalink to this heading"></a></h2>
<p>LIRICAL provides three commands for receiving phenotype and genotype inputs via CLI, as a phenopacket, or as a YAML file.</p>
<section id="prioritize-run-lirical-with-via-cli-options">
<h3><code class="docutils literal notranslate"><span class="pre">prioritize</span></code> - run LIRICAL with via CLI options<a class="headerlink" href="#prioritize-run-lirical-with-via-cli-options" title="Permalink to this heading"></a></h3>
<p>Since <cite>v2</cite> release, all required inputs can be provided as command line arguments of the <code class="docutils literal notranslate"><span class="pre">prioritize</span></code> command.
This leads to a rather lengthy CLI. However, the CLI can be useful e.g. for using with pipeline engines such
as Nextflow or Snakemake.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">prioritize</span></code> command takes the following options:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-p</span> <span class="pre">|</span> <span class="pre">--observed-phenotypes</span></code>: a comma-separated IDs of HPO IDs
that correspond to the phenotype terms observed in the proband.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-n</span> <span class="pre">|</span> <span class="pre">--negated-phenotypes</span></code>: a comma-separated IDs of HPO IDs
that correspond to the phenotype terms negated/excluded in the proband.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--assembly</span></code> genome build, choose from <cite>hg19</cite> or <cite>hg38</cite>, must be provided if <code class="docutils literal notranslate"><span class="pre">--vcf</span></code> is used (default: <code class="docutils literal notranslate"><span class="pre">hg38</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--vcf</span></code>: path to VCF file with exome/genome sequencing results. The file can be compressed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sample-id</span></code>: proband’s identifier (default: <cite>Sample</cite>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--age</span></code>: proband’s age as an ISO8601 duration
(e.g. <code class="docutils literal notranslate"><span class="pre">P9Y</span></code> for 9 years, <code class="docutils literal notranslate"><span class="pre">P2Y3M</span></code> for 2 years and 3 months, or <code class="docutils literal notranslate"><span class="pre">P33W</span></code> for the 33th gestational week).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sex</span></code>: proband’s sex, choose from <cite>MALE</cite>, <cite>FEMALE</cite>, <cite>UNKNOWN</cite> (default: <cite>UNKNOWN</cite>).</p></li>
</ul>
</section>
<section id="phenopacket-run-lirical-with-a-phenopacket">
<h3><code class="docutils literal notranslate"><span class="pre">phenopacket</span></code> - run LIRICAL with a Phenopacket<a class="headerlink" href="#phenopacket-run-lirical-with-a-phenopacket" title="Permalink to this heading"></a></h3>
<p>LIRICAL can be run with clinical data (HPO terms) only or with clinical data and a VCF file representing the
results of gene panel, exome, or genome sequencing. The preferred input format is
<a class="reference external" href="https://phenopacket-schema.readthedocs.io/en/latest/">Phenopackets</a>,
an open standard for sharing disease and phenotype information.
This is a new standard of the <a class="reference external" href="https://www.ga4gh.org/">Global Alliance for Genomics and Health</a> that
links detailed phenotype descriptions with disease, patient, and genetic information.</p>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/hpo-textmining.png"><img alt="PhenopacketGenerator" src="_images/hpo-textmining.png" style="width: 75%;" /></a>
<figcaption>
<p><span class="caption-text">For convenience, we provide a tool called
<a class="reference external" href="https://github.com/TheJacksonLaboratory/PhenopacketGenerator">PhenopacketGenerator</a> that can be used
to create a Phenopacket with a list of HPO terms and the path to a VCF file with which LIRICAL can be run.</span><a class="headerlink" href="#id1" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<p>LIRICAL can be run with clinical data (HPO terms) only or with clinical data and a VCF file representing the
results of gene panel, exome, or genome sequencing.</p>
<p>Let’s consider an example of an individual with <a class="reference external" href="https://omim.org/entry/101600">Pfeiffer syndrome</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
  <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;pfeiffer-example&quot;</span><span class="p">,</span>
  <span class="s2">&quot;subject&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;example-1&quot;</span>
  <span class="p">},</span>
  <span class="s2">&quot;phenotypicFeatures&quot;</span><span class="p">:</span> <span class="p">[{</span>
    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;HP:0000244&quot;</span><span class="p">,</span>
      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Turribrachycephaly&quot;</span>
    <span class="p">}</span>
  <span class="p">},</span> <span class="p">{</span>
    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;HP:0001363&quot;</span><span class="p">,</span>
      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Craniosynostosis&quot;</span>
    <span class="p">}</span>
  <span class="p">},</span> <span class="p">{</span>
    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;HP:0000453&quot;</span><span class="p">,</span>
      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Choanal atresia&quot;</span>
    <span class="p">}</span>
  <span class="p">},</span> <span class="p">{</span>
    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;HP:0000327&quot;</span><span class="p">,</span>
      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Hypoplasia of the maxilla&quot;</span>
    <span class="p">}</span>
  <span class="p">},</span> <span class="p">{</span>
    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;HP:0000238&quot;</span><span class="p">,</span>
      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Hydrocephalus&quot;</span>
   <span class="p">}</span>
  <span class="p">}],</span>
  <span class="s2">&quot;metaData&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="s2">&quot;createdBy&quot;</span><span class="p">:</span> <span class="s2">&quot;Peter R.&quot;</span><span class="p">,</span>
    <span class="s2">&quot;resources&quot;</span><span class="p">:</span> <span class="p">[{</span>
      <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="s2">&quot;hp&quot;</span><span class="p">,</span>
      <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;human phenotype ontology&quot;</span><span class="p">,</span>
      <span class="s2">&quot;namespacePrefix&quot;</span><span class="p">:</span> <span class="s2">&quot;HP&quot;</span><span class="p">,</span>
      <span class="s2">&quot;url&quot;</span><span class="p">:</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/hp.owl&quot;</span><span class="p">,</span>
      <span class="s2">&quot;version&quot;</span><span class="p">:</span> <span class="s2">&quot;2018-03-08&quot;</span><span class="p">,</span>
      <span class="s2">&quot;iriPrefix&quot;</span><span class="p">:</span> <span class="s2">&quot;http://purl.obolibrary.org/obo/HP_&quot;</span>
    <span class="p">}],</span>
    <span class="s2">&quot;phenopacketSchemaVersion&quot;</span><span class="p">:</span> <span class="s2">&quot;2.0.0&quot;</span>
  <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Save the file above as <code class="docutils literal notranslate"><span class="pre">pfeiffer.json</span></code>.</p>
<p><strong>Running LIRICAL with clinical data</strong></p>
<p>LIRICAL will perform phenotype-only analysis if the <code class="docutils literal notranslate"><span class="pre">phenopacket</span></code> command incantation does not contain a <code class="docutils literal notranslate"><span class="pre">--vcf</span></code> option.
In this case, the only required argument is the phenopacket:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="n">phenopacket</span> <span class="o">-</span><span class="n">p</span> <span class="n">pfeiffer</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
<p><strong>Running LIRICAL with a VCF file</strong></p>
<p>Alternatively, LIRICAL can include the VCF file if the path is provided using <code class="docutils literal notranslate"><span class="pre">--vcf</span></code> option.
Note, we must also provide <code class="docutils literal notranslate"><span class="pre">--assembly</span></code> and <code class="docutils literal notranslate"><span class="pre">-e19</span></code> (or <code class="docutils literal notranslate"><span class="pre">-e38</span></code>) options to indicate the genome assembly and path to Exomiser variant database:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="n">phenopacket</span> <span class="o">-</span><span class="n">p</span> <span class="n">pfeiffer</span><span class="o">.</span><span class="n">json</span> <span class="o">--</span><span class="n">vcf</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">pfeiffer</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">assembly</span> <span class="n">hg19</span> <span class="o">-</span><span class="n">e19</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">exomiser</span><span class="o">/</span><span class="mi">2302</span><span class="n">_hg19_variants</span><span class="o">.</span><span class="n">mv</span><span class="o">.</span><span class="n">db</span>
</pre></div>
</div>
</section>
<section id="yaml-running-lirical-with-a-yaml-file">
<h3><code class="docutils literal notranslate"><span class="pre">yaml</span></code> - running LIRICAL with a YAML file<a class="headerlink" href="#yaml-running-lirical-with-a-yaml-file" title="Permalink to this heading"></a></h3>
<p>The other allowed input format is <a class="reference internal" href="yaml.html#rstyaml"><span class="std std-ref">YAML input format</span></a>.</p>
<p>A typical command that runs LIRICAL using settings shown in the YAML file with the default data directory
would be simply:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="n">yaml</span> <span class="o">-</span><span class="n">y</span> <span class="n">example</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
<p>This will run the phenotype-only analysis of the <em>Patient 4</em>.</p>
<p>To run the genotype-aware analysis, modify the YAML file such that the <code class="docutils literal notranslate"><span class="pre">vcf</span></code> field points to the location
of the VCF file on your file system. Then, the analysis is run as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lirical</span> <span class="n">yaml</span> <span class="o">-</span><span class="n">y</span> <span class="n">example</span><span class="o">.</span><span class="n">yml</span> <span class="o">--</span><span class="n">assembly</span> <span class="n">hg19</span> <span class="o">-</span><span class="n">e19</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">exomiser</span><span class="o">/</span><span class="mi">2302</span><span class="n">_hg19_variants</span><span class="o">.</span><span class="n">mv</span><span class="o">.</span><span class="n">db</span>
</pre></div>
</div>
</section>
</section>
<section id="choosing-between-yaml-and-phenopacket-input-formats">
<h2>Choosing between YAML and Phenopacket input formats<a class="headerlink" href="#choosing-between-yaml-and-phenopacket-input-formats" title="Permalink to this heading"></a></h2>
<p>How should users choose between <a class="reference internal" href="yaml-or-phenopacket.html#rstyamlorphenopackethpo"><span class="std std-ref">YAML or Phenopacket as input</span></a>?</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorial.html" class="btn btn-neutral float-left" title="Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="output.html" class="btn btn-neutral float-right" title="Output files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Peter Robinson.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>